From: Daniel van Vugt <daniel.van.vugt@canonical.com>
Date: Thu, 5 Feb 2026 16:14:40 +0800
Subject: Unblock Nvidia performance

Reorganises the "native" (Wayland) backend to accomodate the Nvidia
driver's tendency to block while rendering completes, so that Mutter's
main loop is no longer blocked during that time. This reduces time
spent blocked each frame from milliseconds to microseconds, resulting
in noticeably improved smoothness.

For single GPU configurations, this is 100% feature complete. Dual
GPU configurations however will only see a partial improvement until
such time as upstream rewrites the secondary GPU logic in some future
release 51+. So that's why the Ubuntu bug is declared fixed here, but
not the upstream bug.

Origin: https://gitlab.gnome.org/GNOME/mutter/-/merge_requests/4868
Bug-Ubuntu: https://bugs.launchpad.net/bugs/2081140
Related: https://gitlab.gnome.org/GNOME/mutter/-/issues/3461
---
 src/backends/native/meta-frame-native.c            |  47 +-
 src/backends/native/meta-frame-native.h            |  11 +
 src/backends/native/meta-onscreen-native.c         | 588 ++++++++++++++++-----
 src/backends/native/meta-renderer-native-private.h |   2 +-
 src/backends/native/meta-renderer-native.c         |   2 +-
 5 files changed, 512 insertions(+), 138 deletions(-)

diff --git a/src/backends/native/meta-frame-native.c b/src/backends/native/meta-frame-native.c
index 98cd917..7699863 100644
--- a/src/backends/native/meta-frame-native.c
+++ b/src/backends/native/meta-frame-native.c
@@ -35,7 +35,8 @@ struct _MetaFrameNative
   MetaKmsUpdate *kms_update;
 
   MtkRegion *damage;
-  int sync_fd;
+
+  GPollFD sync;
 };
 
 static void
@@ -43,7 +44,7 @@ meta_frame_native_release (ClutterFrame *frame)
 {
   MetaFrameNative *frame_native = meta_frame_native_from_frame (frame);
 
-  g_clear_fd (&frame_native->sync_fd, NULL);
+  g_clear_fd (&frame_native->sync.fd, NULL);
   g_clear_pointer (&frame_native->damage, mtk_region_unref);
   g_clear_object (&frame_native->buffer);
   g_clear_object (&frame_native->scanout);
@@ -57,7 +58,8 @@ meta_frame_native_new (void)
   MetaFrameNative *frame_native =
     clutter_frame_new (MetaFrameNative, meta_frame_native_release);
 
-  frame_native->sync_fd = -1;
+  frame_native->sync.fd = -1;
+  frame_native->sync.events = 0;
 
   return frame_native;
 }
@@ -139,12 +141,45 @@ void
 meta_frame_native_set_sync_fd (MetaFrameNative *frame_native,
                                int              sync_fd)
 {
-  g_clear_fd (&frame_native->sync_fd, NULL);
-  frame_native->sync_fd = sync_fd;
+  g_clear_fd (&frame_native->sync.fd, NULL);
+  frame_native->sync.fd = sync_fd;
 }
 
 int
 meta_frame_native_steal_sync_fd (MetaFrameNative *frame_native)
 {
-  return g_steal_fd (&frame_native->sync_fd);
+  g_warn_if_fail (frame_native->sync.events == 0);
+
+  return g_steal_fd (&frame_native->sync.fd);
+}
+
+void
+meta_frame_native_add_source (MetaFrameNative *frame_native,
+                              GSource         *source)
+{
+  g_return_if_fail (frame_native->sync.fd >= 0);
+  g_return_if_fail (frame_native->sync.events == 0);
+
+  frame_native->sync.events = G_IO_IN;
+  g_source_add_poll (source, &frame_native->sync);
+}
+
+void
+meta_frame_native_remove_source (MetaFrameNative *frame_native,
+                                 GSource         *source)
+{
+  g_return_if_fail (frame_native->sync.fd >= 0);
+  g_return_if_fail (frame_native->sync.events != 0);
+
+  g_source_remove_poll (source, &frame_native->sync);
+  frame_native->sync.events = 0;
+}
+
+gboolean
+meta_frame_native_is_ready (MetaFrameNative *frame_native)
+{
+  g_return_val_if_fail (frame_native->sync.fd >= 0, FALSE);
+  g_return_val_if_fail (frame_native->sync.events != 0, FALSE);
+
+  return !!(frame_native->sync.revents & G_IO_IN);
 }
diff --git a/src/backends/native/meta-frame-native.h b/src/backends/native/meta-frame-native.h
index 38547eb..cf736da 100644
--- a/src/backends/native/meta-frame-native.h
+++ b/src/backends/native/meta-frame-native.h
@@ -61,3 +61,14 @@ meta_frame_native_set_sync_fd (MetaFrameNative *frame_native,
 
 int
 meta_frame_native_steal_sync_fd (MetaFrameNative *frame_native);
+
+void
+meta_frame_native_add_source (MetaFrameNative *frame_native,
+                              GSource         *source);
+
+void
+meta_frame_native_remove_source (MetaFrameNative *frame_native,
+                                 GSource         *source);
+
+gboolean
+meta_frame_native_is_ready (MetaFrameNative *frame_native);
diff --git a/src/backends/native/meta-onscreen-native.c b/src/backends/native/meta-onscreen-native.c
index 0748e62..5ec0d0e 100644
--- a/src/backends/native/meta-onscreen-native.c
+++ b/src/backends/native/meta-onscreen-native.c
@@ -104,8 +104,16 @@ typedef struct _MetaOnscreenNativeSecondaryGpuState
   gboolean noted_primary_gpu_copy_ok;
   gboolean noted_primary_gpu_copy_failed;
   MetaSharedFramebufferImportStatus import_status;
+  MetaDrmBuffer *source_framebuffer;
 } MetaOnscreenNativeSecondaryGpuState;
 
+typedef struct _RenderSource
+{
+  GSource source;
+  MetaOnscreenNative *onscreen_native;
+  GHashTable *frames;
+} RenderSource;
+
 typedef struct _KmsProperty
 {
   gboolean invalidated;
@@ -148,6 +156,8 @@ struct _MetaOnscreenNative
 
   MetaRendererView *view;
 
+  GSource *render_source;
+
   union {
     struct {
       KmsProperty gamma_lut;
@@ -175,6 +185,194 @@ init_secondary_gpu_state (MetaRendererNative  *renderer_native,
                           CoglOnscreen        *onscreen,
                           GError             **error);
 
+static MetaEgl *
+meta_onscreen_native_get_egl (MetaOnscreenNative *onscreen_native);
+
+static void
+render_source_remove_frame (GSource      *source,
+                            ClutterFrame *frame)
+{
+  RenderSource *render_source;
+
+  if (!source || !frame)
+    return;
+
+  render_source = (RenderSource *) source;
+  if (g_hash_table_contains (render_source->frames, frame))
+    {
+      MetaFrameNative *frame_native = meta_frame_native_from_frame (frame);
+
+      meta_frame_native_remove_source (frame_native, source);
+      g_hash_table_remove (render_source->frames, frame);
+    }
+}
+
+static void
+render_source_add_frame (GSource      *source,
+                         ClutterFrame *frame)
+{
+  RenderSource *render_source;
+  MetaFrameNative *frame_native;
+
+  g_return_if_fail (source != NULL);
+  g_return_if_fail (frame != NULL);
+
+  render_source = (RenderSource *) source;
+  g_return_if_fail (!g_hash_table_contains (render_source->frames, frame));
+
+  frame_native = meta_frame_native_from_frame (frame);
+  meta_frame_native_add_source (frame_native, source);
+  g_hash_table_insert (render_source->frames, clutter_frame_ref (frame), NULL);
+}
+
+static gboolean
+render_source_ready (GSource *source)
+{
+  RenderSource *render_source = (RenderSource *) source;
+  MetaOnscreenNative *onscreen_native = render_source->onscreen_native;
+  ClutterFrame *frame = onscreen_native->next_frame;
+  MetaFrameNative *frame_native;
+
+  if (frame == NULL)
+    return FALSE;
+
+  if (!g_hash_table_contains (render_source->frames, frame))
+    return FALSE;
+
+  frame_native = meta_frame_native_from_frame (frame);
+  return meta_frame_native_is_ready (frame_native);
+}
+
+static void
+maybe_post_next_frame_if_gl_finished (CoglOnscreen *onscreen)
+{
+  MetaOnscreenNative *onscreen_native = META_ONSCREEN_NATIVE (onscreen);
+  RenderSource *render_source = (RenderSource *) onscreen_native->render_source;
+  ClutterFrame *frame = onscreen_native->next_frame;
+
+  if (frame == NULL)
+    return;
+
+  if (render_source &&
+      g_hash_table_contains (render_source->frames, frame) &&
+      !meta_frame_native_is_ready (meta_frame_native_from_frame (frame)))
+    {
+      return;
+    }
+
+  maybe_post_next_frame (onscreen);
+}
+
+static gboolean
+render_source_prepare (GSource *source,
+                       gint    *timeout_ms)
+{
+  *timeout_ms = -1;
+
+  return render_source_ready (source);
+}
+
+static gboolean
+render_source_check (GSource *source)
+{
+  return render_source_ready (source);
+}
+
+static gboolean
+render_source_dispatch (GSource     *source,
+                        GSourceFunc  callback,
+                        gpointer     user_data)
+{
+  if (callback)
+    callback (user_data);
+
+  return G_SOURCE_CONTINUE;
+}
+
+static void
+render_source_finalize (GSource *source)
+{
+  RenderSource *render_source = (RenderSource *) source;
+
+  g_clear_pointer (&render_source->frames, g_hash_table_destroy);
+}
+
+static void
+maybe_init_render_source (MetaOnscreenNative *onscreen_native)
+{
+  CoglFramebuffer *framebuffer = COGL_FRAMEBUFFER (onscreen_native);
+  CoglContext *cogl_context = cogl_framebuffer_get_context (framebuffer);
+  const char *force_render_source;
+  gboolean use_render_source;
+
+  force_render_source = g_getenv ("MUTTER_DEBUG_FORCE_RENDER_SOURCE");
+  if (force_render_source != NULL)
+    {
+      use_render_source = g_strcmp0 (force_render_source, "0") != 0;
+    }
+  else
+    {
+      MetaOnscreenNativeSecondaryGpuState *secondary_gpu_state =
+        onscreen_native->secondary_gpu_state;
+      gboolean output_gpu_is_nvidia;
+
+      if (secondary_gpu_state)
+        {
+          MetaRendererNativeGpuData *renderer_gpu_data =
+            secondary_gpu_state->renderer_gpu_data;
+
+          output_gpu_is_nvidia = renderer_gpu_data->secondary.is_nvidia;
+        }
+      else
+        {
+          MetaEgl *egl = meta_onscreen_native_get_egl (onscreen_native);
+          EGLDisplay egl_display = cogl_context_get_egl_display (cogl_context);
+          const char *egl_vendor =
+            meta_egl_query_string (egl, egl_display, EGL_VENDOR);
+          output_gpu_is_nvidia =
+            g_strcmp0 (egl_vendor, "NVIDIA") == 0;
+        }
+
+      use_render_source = output_gpu_is_nvidia;
+    }
+
+  if (use_render_source)
+    {
+      static GSourceFuncs render_source_funcs = {
+        .prepare = render_source_prepare,
+        .check = render_source_check,
+        .dispatch = render_source_dispatch,
+        .finalize = render_source_finalize,
+      };
+      GSource *source;
+      RenderSource *render_source;
+
+      if (!cogl_context_has_winsys_feature (cogl_context,
+                                            COGL_WINSYS_FEATURE_SYNC_FD))
+        {
+          g_warning ("Render source feature was requested but is disabled "
+                     "due to lack of driver support.");
+          return;
+        }
+
+      source = g_source_new (&render_source_funcs, sizeof (RenderSource));
+      onscreen_native->render_source = source;
+      render_source = (RenderSource *) source;
+      render_source->frames = g_hash_table_new_full (g_direct_hash,
+                                                     g_direct_equal,
+                                                     (GDestroyNotify) clutter_frame_unref,
+                                                     NULL);
+      render_source->onscreen_native = onscreen_native;
+      g_source_set_name (source, "MetaOnscreenNative.render_source");
+      g_source_set_can_recurse (source, FALSE);
+      g_source_set_callback (source,
+                             (GSourceFunc) maybe_post_next_frame_if_gl_finished,
+                             onscreen_native,
+                             NULL);
+      g_source_attach (source, NULL);
+    }
+}
+
 static void
 meta_onscreen_native_promote_posted_frame (CoglOnscreen *onscreen)
 {
@@ -282,7 +480,7 @@ notify_view_crtc_presented (MetaRendererView *view,
 
   meta_onscreen_native_notify_frame_complete (onscreen);
   meta_onscreen_native_promote_posted_frame (onscreen);
-  maybe_post_next_frame (onscreen);
+  maybe_post_next_frame_if_gl_finished (onscreen);
 }
 
 static void
@@ -339,7 +537,7 @@ page_flip_feedback_ready (MetaKmsCrtc *kms_crtc,
 
   meta_onscreen_native_notify_frame_complete (onscreen);
   meta_onscreen_native_promote_posted_frame (onscreen);
-  maybe_post_next_frame (onscreen);
+  maybe_post_next_frame_if_gl_finished (onscreen);
 }
 
 static void
@@ -407,7 +605,7 @@ page_flip_feedback_discarded (MetaKmsCrtc  *kms_crtc,
 
   meta_onscreen_native_notify_frame_complete (onscreen);
   meta_onscreen_native_clear_posted_fb (onscreen);
-  maybe_post_next_frame (onscreen);
+  maybe_post_next_frame_if_gl_finished (onscreen);
 }
 
 static const MetaKmsPageFlipListenerVtable page_flip_listener_vtable = {
@@ -878,6 +1076,8 @@ secondary_gpu_state_free (MetaOnscreenNativeSecondaryGpuState *secondary_gpu_sta
   MetaBackend *backend = meta_gpu_get_backend (gpu);
   MetaEgl *egl = meta_backend_get_egl (backend);
 
+  g_clear_object (&secondary_gpu_state->source_framebuffer);
+
   if (secondary_gpu_state->egl_surface != EGL_NO_SURFACE)
     {
       MetaRendererNativeGpuData *renderer_gpu_data;
@@ -906,15 +1106,51 @@ secondary_gpu_state_free (MetaOnscreenNativeSecondaryGpuState *secondary_gpu_sta
   g_free (secondary_gpu_state);
 }
 
+static MetaDrmBuffer *
+acquire_primary_front_buffer (CoglOnscreen  *onscreen,
+                              GError       **error)
+{
+  MetaOnscreenNative *onscreen_native = META_ONSCREEN_NATIVE (onscreen);
+  MetaRendererNative *renderer_native = onscreen_native->renderer_native;
+  CoglFramebuffer *framebuffer = COGL_FRAMEBUFFER (onscreen);
+  CoglContext *cogl_context = cogl_framebuffer_get_context (framebuffer);
+  CoglRenderer *cogl_renderer = cogl_context->display->renderer;
+  CoglRendererEGL *cogl_renderer_egl = cogl_renderer_get_winsys_data (cogl_renderer);
+  MetaRendererNativeGpuData *renderer_gpu_data = cogl_renderer_egl->platform;
+  MetaDrmBufferFlags buffer_flags;
+  MetaDrmBufferGbm *buffer_gbm;
+  MetaRenderDevice *render_device = renderer_gpu_data->render_device;
+  MetaDeviceFile *render_device_file =
+    meta_render_device_get_device_file (render_device);
+
+  buffer_flags = META_DRM_BUFFER_FLAG_NONE;
+  if (!meta_renderer_native_use_modifiers (renderer_native))
+    buffer_flags |= META_DRM_BUFFER_FLAG_DISABLE_MODIFIERS;
+
+  buffer_gbm =
+    meta_drm_buffer_gbm_new_lock_front (render_device_file,
+                                        onscreen_native->gbm.surface,
+                                        buffer_flags,
+                                        error);
+  if (!buffer_gbm)
+    return NULL;
+
+  return META_DRM_BUFFER (g_steal_pointer (&buffer_gbm));
+}
+
 static MetaDrmBuffer *
 import_shared_framebuffer (CoglOnscreen                        *onscreen,
-                           MetaOnscreenNativeSecondaryGpuState *secondary_gpu_state,
-                           MetaDrmBuffer                       *primary_gpu_fb)
+                           MetaOnscreenNativeSecondaryGpuState *secondary_gpu_state)
 {
   MetaRenderDevice *render_device;
   g_autoptr (GError) error = NULL;
+  g_autoptr (MetaDrmBuffer) primary_gpu_fb = NULL;
   MetaDrmBuffer *imported_buffer;
 
+  primary_gpu_fb = acquire_primary_front_buffer (onscreen, &error);
+  if (!primary_gpu_fb)
+    return NULL;
+
   render_device = secondary_gpu_state->renderer_gpu_data->render_device;
   imported_buffer =
     meta_render_device_import_dma_buf (render_device,
@@ -1048,14 +1284,13 @@ get_secondary_gpu_buffer_age (MetaOnscreenNativeSecondaryGpuState *secondary_gpu
   return buffer_age;
 }
 
-static MetaDrmBuffer *
-copy_shared_framebuffer_gpu (CoglOnscreen                         *onscreen,
-                             MetaFrameNative                      *frame_native,
-                             MetaOnscreenNativeSecondaryGpuState  *secondary_gpu_state,
-                             MetaRendererNativeGpuData            *renderer_gpu_data,
-                             MetaDrmBuffer                        *primary_gpu_fb,
-                             const MtkRegion                      *region,
-                             GError                              **error)
+static void
+start_copy_shared_framebuffer_gpu (CoglOnscreen                         *onscreen,
+                                   int                                  *sync_fd,
+                                   MetaOnscreenNativeSecondaryGpuState  *secondary_gpu_state,
+                                   MetaRendererNativeGpuData            *renderer_gpu_data,
+                                   const MtkRegion                      *region,
+                                   GError                              **error)
 {
   MetaRendererNative *renderer_native = renderer_gpu_data->renderer_native;
   MetaEgl *egl = meta_renderer_native_get_egl (renderer_native);
@@ -1065,23 +1300,24 @@ copy_shared_framebuffer_gpu (CoglOnscreen                         *onscreen,
   CoglDisplay *cogl_display = cogl_context_get_display (cogl_context);
   MetaRenderDevice *render_device;
   EGLDisplay egl_display;
-  gboolean use_modifiers;
-  MetaDeviceFile *device_file;
-  MetaDrmBufferFlags flags;
   MetaDrmBufferGbm *buffer_gbm = NULL;
   struct gbm_bo *bo;
   EGLSync egl_sync = EGL_NO_SYNC;
-  g_autofd int sync_fd = -1;
   EGLImageKHR egl_image;
   int buffer_age = 0;
   g_autoptr (MtkRegion) blit_region = NULL;
   g_autoptr (MtkRegion) region_to_push = NULL;
+  g_autoptr (MetaDrmBuffer) primary_gpu_fb = NULL;
 
-  COGL_TRACE_BEGIN_SCOPED (CopySharedFramebufferSecondaryGpu,
-                           "copy_shared_framebuffer_gpu()");
+  COGL_TRACE_BEGIN_SCOPED (StartCopySharedFramebufferSecondaryGpu,
+                           "start_copy_shared_framebuffer_gpu()");
+
+  primary_gpu_fb = acquire_primary_front_buffer (onscreen, error);
+  if (!primary_gpu_fb)
+    return;
 
-  if (renderer_gpu_data->secondary.needs_explicit_sync)
-    sync_fd = meta_frame_native_steal_sync_fd (frame_native);
+  if (!renderer_gpu_data->secondary.is_nvidia)
+    g_clear_fd (sync_fd, NULL);
 
   render_device = renderer_gpu_data->render_device;
   egl_display = meta_render_device_get_egl_display (render_device);
@@ -1097,12 +1333,12 @@ copy_shared_framebuffer_gpu (CoglOnscreen                         *onscreen,
       goto done;
     }
 
-  if (sync_fd >= 0)
+  if (*sync_fd >= 0)
     {
       EGLAttrib attribs[3];
 
       attribs[0] = EGL_SYNC_NATIVE_FENCE_FD_ANDROID;
-      attribs[1] = g_steal_fd (&sync_fd);
+      attribs[1] = g_steal_fd (sync_fd);
       attribs[2] = EGL_NONE;
 
       if (!meta_egl_create_sync (egl,
@@ -1166,14 +1402,44 @@ copy_shared_framebuffer_gpu (CoglOnscreen                         *onscreen,
       goto done;
     }
 
-  sync_fd = meta_egl_create_sync_fd (egl, egl_display, error);
-  if (sync_fd < 0)
+  *sync_fd = meta_egl_create_sync_fd (egl, egl_display, error);
+  if (*sync_fd < 0)
     {
       g_prefix_error (error, "Failed to create sync fd: ");
       goto done;
     }
 
-  meta_frame_native_set_sync_fd (frame_native, g_steal_fd (&sync_fd));
+done:
+  if (egl_sync != EGL_NO_SYNC)
+    {
+      g_autoptr (GError) local_error = NULL;
+
+      if (!meta_egl_destroy_sync (egl,
+                                  egl_display,
+                                  egl_sync,
+                                  &local_error))
+        g_warning ("Failed to destroy secondary GPU EGLSync: %s", local_error->message);
+    }
+
+  _cogl_winsys_egl_ensure_current (cogl_display);
+}
+
+static MetaDrmBuffer *
+finish_copy_shared_framebuffer_gpu (CoglOnscreen                         *onscreen,
+                                    MetaFrameNative                      *frame_native,
+                                    MetaOnscreenNativeSecondaryGpuState  *secondary_gpu_state,
+                                    MetaRendererNativeGpuData            *renderer_gpu_data,
+                                    GError                              **error)
+{
+  MetaRendererNative *renderer_native = renderer_gpu_data->renderer_native;
+  MetaRenderDevice *render_device = renderer_gpu_data->render_device;
+  gboolean use_modifiers;
+  MetaDeviceFile *device_file;
+  MetaDrmBufferFlags flags;
+  MetaDrmBufferGbm *buffer_gbm = NULL;
+
+  COGL_TRACE_BEGIN_SCOPED (FinishCopySharedFramebufferSecondaryGpu,
+                           "finish_copy_shared_framebuffer_gpu()");
 
   use_modifiers = meta_renderer_native_use_modifiers (renderer_native);
   device_file = meta_render_device_get_device_file (render_device);
@@ -1188,29 +1454,7 @@ copy_shared_framebuffer_gpu (CoglOnscreen                         *onscreen,
                                         flags,
                                         error);
   if (!buffer_gbm)
-    {
-      g_prefix_error (error, "meta_drm_buffer_gbm_new_lock_front failed: ");
-      goto done;
-    }
-
-  g_object_set_qdata_full (G_OBJECT (buffer_gbm),
-                           blit_source_quark,
-                           g_object_ref (primary_gpu_fb),
-                           g_object_unref);
-
-done:
-  if (egl_sync != EGL_NO_SYNC)
-    {
-      g_autoptr (GError) local_error = NULL;
-
-      if (!meta_egl_destroy_sync (egl,
-                                  egl_display,
-                                  egl_sync,
-                                  &local_error))
-        g_warning ("Failed to destroy secondary GPU EGLSync: %s", local_error->message);
-    }
-
-  _cogl_winsys_egl_ensure_current (cogl_display);
+    g_prefix_error (error, "meta_drm_buffer_gbm_new_lock_front failed: ");
 
   return buffer_gbm ? META_DRM_BUFFER (buffer_gbm) : NULL;
 }
@@ -1478,12 +1722,38 @@ update_secondary_gpu_state_pre_swap_buffers (CoglOnscreen    *onscreen,
   return copy;
 }
 
+static void
+update_sync_fd_post_swap_buffers (int              *sync_fd,
+                                  CoglOnscreen     *onscreen,
+                                  const MtkRegion  *region,
+                                  GError          **error)
+{
+  MetaOnscreenNative *onscreen_native = META_ONSCREEN_NATIVE (onscreen);
+  MetaOnscreenNativeSecondaryGpuState *secondary_gpu_state =
+    onscreen_native->secondary_gpu_state;
+
+  if (secondary_gpu_state)
+    {
+      MetaRendererNativeGpuData *renderer_gpu_data =
+        secondary_gpu_state->renderer_gpu_data;
+
+      if (renderer_gpu_data->secondary.copy_mode ==
+          META_SHARED_FRAMEBUFFER_COPY_MODE_SECONDARY_GPU)
+        {
+          start_copy_shared_framebuffer_gpu (onscreen,
+                                             sync_fd,
+                                             secondary_gpu_state,
+                                             renderer_gpu_data,
+                                             region,
+                                             error);
+        }
+    }
+}
+
 static MetaDrmBuffer *
 acquire_front_buffer (CoglOnscreen     *onscreen,
                       MetaFrameNative  *frame_native,
-                      MetaDrmBuffer    *primary_gpu_fb,
                       MetaDrmBuffer    *secondary_gpu_fb,
-                      const MtkRegion  *region,
                       GError          **error)
 {
   MetaOnscreenNative *onscreen_native = META_ONSCREEN_NATIVE (onscreen);
@@ -1497,7 +1767,7 @@ acquire_front_buffer (CoglOnscreen     *onscreen,
 
   secondary_gpu_state = onscreen_native->secondary_gpu_state;
   if (!secondary_gpu_state)
-    return g_object_ref (primary_gpu_fb);
+    return acquire_primary_front_buffer (onscreen, error);
 
   renderer_gpu_data =
     meta_renderer_native_get_gpu_data (renderer_native,
@@ -1506,8 +1776,7 @@ acquire_front_buffer (CoglOnscreen     *onscreen,
     {
     case META_SHARED_FRAMEBUFFER_COPY_MODE_ZERO:
       imported_fb = import_shared_framebuffer (onscreen,
-                                               secondary_gpu_state,
-                                               primary_gpu_fb);
+                                               secondary_gpu_state);
       if (imported_fb)
         return imported_fb;
       /* The fallback was prepared in pre_swap_buffers and is currently
@@ -1525,13 +1794,11 @@ acquire_front_buffer (CoglOnscreen     *onscreen,
         }
       return g_object_ref (secondary_gpu_fb);
     case META_SHARED_FRAMEBUFFER_COPY_MODE_SECONDARY_GPU:
-      return copy_shared_framebuffer_gpu (onscreen,
-                                          frame_native,
-                                          secondary_gpu_state,
-                                          renderer_gpu_data,
-                                          primary_gpu_fb,
-                                          region,
-                                          error);
+      return finish_copy_shared_framebuffer_gpu (onscreen,
+                                                 frame_native,
+                                                 secondary_gpu_state,
+                                                 renderer_gpu_data,
+                                                 error);
     }
 
   g_assert_not_reached ();
@@ -1610,6 +1877,8 @@ assign_next_frame (MetaOnscreenNative *onscreen_native,
 
   if (onscreen_native->next_frame != NULL)
     {
+      render_source_remove_frame (onscreen_native->render_source,
+                                  onscreen_native->next_frame);
       clear_superseded_frame (onscreen);
       onscreen_native->superseded_frame =
         g_steal_pointer (&onscreen_native->next_frame);
@@ -1638,30 +1907,49 @@ meta_onscreen_native_swap_buffers_with_damage (CoglOnscreen    *onscreen,
 {
   CoglFramebuffer *framebuffer = COGL_FRAMEBUFFER (onscreen);
   CoglContext *cogl_context = cogl_framebuffer_get_context (framebuffer);
-  CoglRenderer *cogl_renderer = cogl_context->display->renderer;
-  CoglRendererEGL *cogl_renderer_egl = cogl_renderer_get_winsys_data (cogl_renderer);
-  MetaRendererNativeGpuData *renderer_gpu_data = cogl_renderer_egl->platform;
-  MetaRendererNative *renderer_native = renderer_gpu_data->renderer_native;
   MetaOnscreenNative *onscreen_native = META_ONSCREEN_NATIVE (onscreen);
-  MetaGpuKms *render_gpu = onscreen_native->render_gpu;
-  MetaDeviceFile *render_device_file;
+  MetaOnscreenNativeSecondaryGpuState *secondary_gpu_state;
   ClutterFrame *frame = user_data;
   MetaFrameNative *frame_native = meta_frame_native_from_frame (frame);
   CoglOnscreenClass *parent_class;
   g_autoptr (GError) error = NULL;
-  MetaDrmBufferFlags buffer_flags;
-  MetaDrmBufferGbm *buffer_gbm;
-  g_autoptr (MetaDrmBuffer) primary_gpu_fb = NULL;
   g_autoptr (MetaDrmBuffer) secondary_gpu_fb = NULL;
-  g_autoptr (MetaDrmBuffer) buffer = NULL;
   g_autofd int sync_fd = -1;
 
   COGL_TRACE_BEGIN_SCOPED (MetaRendererNativeSwapBuffers,
                            "Meta::OnscreenNative::swap_buffers_with_damage()");
 
+  if (onscreen_native->next_frame != NULL)
+    {
+      /* This is the slow path (though still faster than what came before it)
+       * because it bypasses the render_source. It's here mainly because
+       * Nvidia's eglSwapBuffers will deadlock unless paired perfectly
+       * with the same number of gbm_surface_lock_front_buffer calls.
+       * Mesa does not seem to have this issue and allows more swaps than
+       * locks (has more buffers?).
+       *
+       * Disabling the render source, or disabling triple buffering avoids
+       * this path, but also results in noticeably lower performance on
+       * Nvidia.
+       */
+      maybe_post_next_frame (onscreen);
+
+      g_warn_if_fail (onscreen_native->next_frame == NULL);
+      if (onscreen_native->next_frame != NULL)
+        goto swap_failed;
+    }
+
   secondary_gpu_fb =
     update_secondary_gpu_state_pre_swap_buffers (onscreen, region);
 
+  secondary_gpu_state = onscreen_native->secondary_gpu_state;
+  if (secondary_gpu_state)
+    {
+      g_clear_object (&secondary_gpu_state->source_framebuffer);
+      secondary_gpu_state->source_framebuffer =
+        g_steal_pointer (&secondary_gpu_fb);
+    }
+
   parent_class = COGL_ONSCREEN_CLASS (meta_onscreen_native_parent_class);
   parent_class->swap_buffers_with_damage (onscreen,
                                           region,
@@ -1669,62 +1957,11 @@ meta_onscreen_native_swap_buffers_with_damage (CoglOnscreen    *onscreen,
                                           user_data);
 
   sync_fd = cogl_context_get_latest_sync_fd (cogl_context);
-  if (sync_fd >= 0)
-    meta_frame_native_set_sync_fd (frame_native, g_steal_fd (&sync_fd));
-
-  renderer_gpu_data = meta_renderer_native_get_gpu_data (renderer_native,
-                                                         render_gpu);
-  render_device_file =
-    meta_render_device_get_device_file (renderer_gpu_data->render_device);
-  switch (renderer_gpu_data->mode)
+  update_sync_fd_post_swap_buffers (&sync_fd, onscreen, region, &error);
+  if (sync_fd < 0 && error != NULL)
     {
-    case META_RENDERER_NATIVE_MODE_GBM:
-      buffer_flags = META_DRM_BUFFER_FLAG_NONE;
-      if (!meta_renderer_native_use_modifiers (renderer_native))
-        buffer_flags |= META_DRM_BUFFER_FLAG_DISABLE_MODIFIERS;
-
-      buffer_gbm =
-        meta_drm_buffer_gbm_new_lock_front (render_device_file,
-                                            onscreen_native->gbm.surface,
-                                            buffer_flags,
-                                            &error);
-      if (!buffer_gbm)
-        {
-          g_warning ("Failed to lock front buffer on %s: %s",
-                     meta_device_file_get_path (render_device_file),
-                     error->message);
-          goto swap_failed;
-        }
-
-      primary_gpu_fb = META_DRM_BUFFER (g_steal_pointer (&buffer_gbm));
-      buffer = acquire_front_buffer (onscreen,
-                                     frame_native,
-                                     primary_gpu_fb,
-                                     secondary_gpu_fb,
-                                     region,
-                                     &error);
-      if (buffer == NULL)
-        {
-          g_warning ("Failed to acquire front buffer: %s", error->message);
-          goto swap_failed;
-        }
-
-      meta_frame_native_set_buffer (frame_native, buffer);
-
-      if (!meta_drm_buffer_ensure_fb_id (buffer, &error))
-        {
-          g_warning ("Failed to ensure KMS FB ID on %s: %s",
-                     meta_device_file_get_path (render_device_file),
-                     error->message);
-          goto swap_failed;
-        }
-      break;
-    case META_RENDERER_NATIVE_MODE_SURFACELESS:
-      break;
-#ifdef HAVE_EGL_DEVICE
-    case META_RENDERER_NATIVE_MODE_EGL_DEVICE:
-      break;
-#endif
+      g_warning ("update_sync_fd_post_swap_buffers failed: %s", error->message);
+      goto swap_failed;
     }
 
   assign_next_frame (onscreen_native, frame);
@@ -1734,7 +1971,19 @@ meta_onscreen_native_swap_buffers_with_damage (CoglOnscreen    *onscreen,
 
   meta_frame_native_set_damage (frame_native, region);
 
-  maybe_post_next_frame (onscreen);
+  if (sync_fd >= 0)
+    {
+      meta_frame_native_set_sync_fd (frame_native, g_steal_fd (&sync_fd));
+      if (onscreen_native->render_source)
+        render_source_add_frame (onscreen_native->render_source, frame);
+      else
+        maybe_post_next_frame (onscreen);
+    }
+  else
+    {
+      maybe_post_next_frame (onscreen);
+    }
+
   return;
 
 swap_failed:
@@ -1772,13 +2021,20 @@ maybe_post_next_frame (CoglOnscreen *onscreen)
   const MetaKmsResultListenerVtable *listener;
   MetaKmsAssignPlaneFlag flip_flags;
   gboolean is_direct_scanout;
+  gboolean is_swap_buffers;
   MetaBackendNative *backend_native = META_BACKEND_NATIVE (backend);
   MetaKms *kms = meta_backend_native_get_kms (backend_native);
+  CoglFrameInfo *frame_info;
 
   COGL_TRACE_SCOPED_ANCHOR (MetaRendererNativePostKmsUpdate);
 
-  if (onscreen_native->next_frame == NULL ||
-      onscreen_native->posted_frame != NULL ||
+  if (onscreen_native->next_frame == NULL)
+    return;
+
+  render_source_remove_frame (onscreen_native->render_source,
+                              onscreen_native->next_frame);
+
+  if (onscreen_native->posted_frame != NULL ||
       onscreen_native->view == NULL ||
       meta_kms_is_shutting_down (kms))
     return;
@@ -1791,6 +2047,8 @@ maybe_post_next_frame (CoglOnscreen *onscreen)
       return;
     }
 
+  render_source_remove_frame (onscreen_native->render_source,
+                              onscreen_native->next_frame);
   frame = g_steal_pointer (&onscreen_native->next_frame);
   frame_native = meta_frame_native_from_frame (frame);
   region = meta_frame_native_get_damage (frame_native);
@@ -1803,12 +2061,64 @@ maybe_post_next_frame (CoglOnscreen *onscreen)
                                                     kms_device);
 
   is_direct_scanout = meta_frame_native_get_scanout (frame_native) != NULL;
+  is_swap_buffers = meta_frame_native_get_damage (frame_native) != NULL;
+
   if (is_direct_scanout)
     {
       listener = &scanout_result_listener_vtable;
       flip_flags = META_KMS_ASSIGN_PLANE_FLAG_DISABLE_IMPLICIT_SYNC;
     }
-  else
+  else if (is_swap_buffers)
+    {
+      g_autoptr (MetaDrmBuffer) secondary_gpu_fb = NULL;
+      g_autoptr (MetaDrmBuffer) buffer = NULL;
+      g_autoptr (GError) error = NULL;
+      MetaOnscreenNativeSecondaryGpuState *secondary_gpu_state;
+      MetaDeviceFile *render_device_file =
+        meta_render_device_get_device_file (renderer_gpu_data->render_device);
+
+      listener = &swap_buffer_result_listener_vtable;
+      flip_flags = META_KMS_ASSIGN_PLANE_FLAG_NONE;
+
+      secondary_gpu_state = onscreen_native->secondary_gpu_state;
+      if (secondary_gpu_state)
+        {
+          secondary_gpu_fb =
+            g_steal_pointer (&secondary_gpu_state->source_framebuffer);
+        }
+
+      switch (renderer_gpu_data->mode)
+        {
+        case META_RENDERER_NATIVE_MODE_GBM:
+          buffer = acquire_front_buffer (onscreen,
+                                         frame_native,
+                                         secondary_gpu_fb,
+                                         &error);
+          if (buffer == NULL)
+            {
+              g_warning ("Failed to acquire front buffer: %s", error->message);
+              goto post_failed;
+            }
+
+          meta_frame_native_set_buffer (frame_native, buffer);
+
+          if (!meta_drm_buffer_ensure_fb_id (buffer, &error))
+            {
+              g_warning ("Failed to ensure KMS FB ID on %s: %s",
+                         meta_device_file_get_path (render_device_file),
+                         error->message);
+              goto post_failed;
+            }
+          break;
+        case META_RENDERER_NATIVE_MODE_SURFACELESS:
+          break;
+#ifdef HAVE_EGL_DEVICE
+        case META_RENDERER_NATIVE_MODE_EGL_DEVICE:
+          break;
+#endif
+        }
+    }
+  else /* property-only update, no framebuffer changes */
     {
       listener = &swap_buffer_result_listener_vtable;
       flip_flags = META_KMS_ASSIGN_PLANE_FLAG_NONE;
@@ -1909,6 +2219,13 @@ maybe_post_next_frame (CoglOnscreen *onscreen)
 
   meta_kms_device_post_update (kms_device, kms_update,
                                META_KMS_UPDATE_FLAG_NONE);
+
+  return;
+
+post_failed:
+  frame_info = cogl_onscreen_peek_head_frame_info (onscreen);
+  frame_info->flags |= COGL_FRAME_INFO_FLAG_SYMBOLIC;
+  meta_onscreen_native_notify_frame_complete (onscreen);
 }
 
 gboolean
@@ -2363,8 +2680,16 @@ void
 meta_onscreen_native_discard_pending_swaps (CoglOnscreen *onscreen)
 {
   MetaOnscreenNative *onscreen_native = META_ONSCREEN_NATIVE (onscreen);
+  MetaOnscreenNativeSecondaryGpuState *secondary_gpu_state;
+
+  secondary_gpu_state = onscreen_native->secondary_gpu_state;
+  if (secondary_gpu_state)
+    g_clear_object (&secondary_gpu_state->source_framebuffer);
 
   discard_pending_swap (&onscreen_native->superseded_frame);
+
+  render_source_remove_frame (onscreen_native->render_source,
+                              onscreen_native->next_frame);
   discard_pending_swap (&onscreen_native->next_frame);
 }
 
@@ -2789,6 +3114,8 @@ meta_onscreen_native_allocate (CoglFramebuffer  *framebuffer,
         return FALSE;
     }
 
+  maybe_init_render_source (onscreen_native);
+
   width = cogl_framebuffer_get_width (framebuffer);
   height = cogl_framebuffer_get_height (framebuffer);
 
@@ -3198,6 +3525,7 @@ meta_onscreen_native_dispose (GObject *object)
   MetaRendererNative *renderer_native = onscreen_native->renderer_native;
   MetaRendererNativeGpuData *renderer_gpu_data;
 
+  g_clear_pointer (&onscreen_native->render_source, g_source_destroy);
   meta_onscreen_native_detach (onscreen_native);
 
   meta_onscreen_native_discard_pending_swaps (onscreen);
diff --git a/src/backends/native/meta-renderer-native-private.h b/src/backends/native/meta-renderer-native-private.h
index 9c7887a..3f94c30 100644
--- a/src/backends/native/meta-renderer-native-private.h
+++ b/src/backends/native/meta-renderer-native-private.h
@@ -61,7 +61,7 @@ typedef struct _MetaRendererNativeGpuData
     MetaSharedFramebufferCopyMode copy_mode;
     gboolean copy_mode_primary_force_cpu;
     gboolean has_EGL_EXT_image_dma_buf_import_modifiers;
-    gboolean needs_explicit_sync;
+    gboolean is_nvidia;
 
     /* For GPU blit mode */
     EGLContext egl_context;
diff --git a/src/backends/native/meta-renderer-native.c b/src/backends/native/meta-renderer-native.c
index f553b6c..5937bde 100644
--- a/src/backends/native/meta-renderer-native.c
+++ b/src/backends/native/meta-renderer-native.c
@@ -2114,7 +2114,7 @@ init_secondary_gpu_data_gpu (MetaRendererNativeGpuData *renderer_gpu_data,
 
   egl_vendor = meta_egl_query_string (egl, egl_display, EGL_VENDOR);
   if (!g_strcmp0 (egl_vendor, "NVIDIA"))
-    renderer_gpu_data->secondary.needs_explicit_sync = TRUE;
+    renderer_gpu_data->secondary.is_nvidia = TRUE;
 
   ret = TRUE;
 out:
